aten/src/ATen/native/cuda/UnaryGammaKernels.cu

```diff
@@ -49,18 +49,53 @@ void trigamma_kernel_cuda(TensorIteratorBase& iter) {
   #endif // USE_JITERATOR
 }
 
-// TODO: jiterate the polygamma kernel (requires supporting heterogenous functor args)
+// Note [polygamma jiterator]
+// To pass a runtime argument (similar to lambda captures in non-JIT kernels),
+// we need to pass to additional arguments to `jitted_gpu_kernel`
+// 1. `vector<pair<string, string>>>` where first string is the
```

Comment:
This approach seems good but I wonder if we can tweak it? A few questions:

- why do we need the argument's name? 
- we should be clear about additional types are supported (just C++ primitive types, basically?) and maybe add an assertion (if it's not too difficult) that we only receive those types
- passing the actual data by address is very general, which is neat, but do we need more functionality than passing primitive C++ types? If not, can we pass those values directly? Would doing so require significant changes to the way we're constructing kernels?
- should the types be a template or runtime parameter? Why or why not?
- why are the additional argument types templates specified by a string and not as actual types?

Reply:
> why do we need the argument's name?

Reading this made we realize that we don't. Removed :)

> we should be clear about additional types are supported (just C++ primitive types, basically?) and maybe add an assertion (if it's not too difficult) that we only receive those types


Have added a static_assert for the invalid type.
<details>

<summary>Sample snippet compiler error on invalid type</summary>

```
../aten/src/ATen/native/cuda/jit_utils.h(64): error: static assertion failed with "invalid type for jiterator"
          detected during:
            instantiation of "std::string at::cuda::jit::typeName<T>() [with T=uint32_t]"
../aten/src/ATen/native/cuda/CUDALoops.cuh(142): here
            instantiation of "c10::SmallVector<std::string, 1U> at::native::<unnamed>::get_extra_args_typenames<Args...>() [with Args=<int64_t, uint32_t>]"
../aten/src/ATen/native/cuda/CUDALoops.cuh(254): here
```


</details>

> passing the actual data by address is very general, which is neat, but do we need more functionality than passing primitive C++ types? If not, can we pass those values directly? Would doing so require significant changes to the way we're constructing kernels?

Turns out this can be done by updating the helper function which helps to convert parameter pack to std::array<void*>

> should the types be a template or runtime parameter? Why or why not?

Thanks for raising the point. Thinking about it, it makes sense to have them as template parameters as we know about them at compile time already. So don't need to pass around a vector of string unnecessarily.

> why are the additional argument types templates specified by a string and not as actual types?

I wasn't sure if we could get the string names of the type without passing a string explicitly (as C++ doesn't have reflection mechanism). But with the review I realized that since the types to be supported were constrained. We can use the `typeName` machinery in `jit_utils.h` for this purpose. (With that it means that all the types that are supported by typeName are valid to be passed in parameter pack.

Thanks for the helpful and thought provoking review :)